src/connector/src/sink/pulsar.rs

```diff
@@ -0,0 +1,321 @@
+            && config.r#type != SINK_TYPE_UPSERT {
+            return Err(SinkError::Config(anyhow!(
+                "`{}` must be {}, or {}",
+                SINK_TYPE_OPTION,
+                SINK_TYPE_APPEND_ONLY,
+                SINK_TYPE_UPSERT
+            )));
+        }
+        Ok(config)
+    }
+}
+
+#[derive(Debug)]
+pub struct PulsarSink {
+    pub config: PulsarConfig,
+    schema: Schema,
+    downstream_pk: Vec<usize>,
+    is_append_only: bool,
+}
+
+impl PulsarSink {
+    pub fn new(config: PulsarConfig, param: SinkParam) -> Self {
+        Self {
+            config,
+            schema: param.schema(),
+            downstream_pk: param.downstream_pk,
+            is_append_only: param.sink_type.is_append_only(),
+        }
+    }
+}
+
+#[async_trait]
+impl Sink for PulsarSink {
+    type Coordinator = DummySinkCommitCoordinator;
+    type Writer = PulsarSinkWriter;
+    async fn new_writer(&self, _writer_param: SinkWriterParam) -> Result<Self::Writer> {
+        PulsarSinkWriter::new(
+            self.config.clone(), 
+            self.schema.clone(), 
+            self.downstream_pk.clone(),
+            self.is_append_only,
+        ).await
+    }
+
+    async fn validate(&self, _client: Option<ConnectorClient>) -> Result<()> {
+        // For upsert Pulsar sink, the primary key must be defined.
+        if !self.is_append_only && self.downstream_pk.is_empty() {
+            return Err(SinkError::Config(anyhow!(
+                "primary key not defined for {} pulsar sink (please define in `primary_key` field)",
+                self.config.r#type
+            )));
+        }
+
+        // TODO: validate pulsar connection
+
+        Ok(())
+    }
+}
+
+pub struct PulsarSinkWriter {
+    pulsar: Pulsar<TokioExecutor>,
+    producer: Producer<TokioExecutor>,
+    config: PulsarConfig,
+    schema: Schema,
+    downstream_pk: Vec<usize>,
+    is_append_only: bool,
+    send_future_buffer: VecDeque<SendFuture>,
+}
+
+impl PulsarSinkWriter {
+    pub async fn new(
+        config: PulsarConfig, 
+        schema: Schema, 
+        downstream_pk: Vec<usize>, 
+        is_append_only: bool
+    ) -> Result<Self> {
+        let pulsar = config.common.build_client().await?;
+        let producer = pulsar
+            .producer()
+            .with_topic(&config.common.topic)
+            .build()
+            .map_err(|e| SinkError::Pulsar(anyhow!("Pulsar sink error: {}", e)))
+            .await?;
+        Ok(Self {
+            pulsar,
+            producer,
+            config,
+            schema,
+            downstream_pk,
+            is_append_only,
+            send_future_buffer: VecDeque::new(),
+        })
+    }
+
+    async fn send_message(&mut self, message: Message) -> Result<()> {
+        let mut success_flag = false;
+        let mut connection_err = None;
+
+        for _ in 0..self.config.max_retry_num {
+            match self.producer.send(message.clone()).await {
```

Comment:
I'm considering if we should have a smaller batch (i.e., 10) and send the messages together using `send_all`.
In this case a `Vec<SendFuture>` will be returned, rather than a single `SendFuture`.
cc @wenym1.

Reply:
Both LGTM. Just check the code of pulsar client. Seems that the implementation of `send_all` is just calling `send` one by one for all messages, so actually there is no extra optimization for batching. If we keep the same logic (i.e. only call `send`), in the future we can extract the common logic and avoid so many duplicated code in kafka and pulsar.

Reply:
```rust
/// sends a list of messages
#[cfg_attr(feature = "telemetry", tracing::instrument(skip_all))]
pub async fn send_all<T, I>(&mut self, messages: I) -> Result<Vec<SendFuture>, Error>
where
    T: SerializeMessage,
    I: IntoIterator<Item = T>,
{
    let producer = match &mut self.inner {
        ProducerInner::Single(p) => p,
        ProducerInner::Partitioned(p) => p.next(),
    };
    let mut sends = Vec::new();
    for message in messages {
        sends.push(producer.send(message).await);
    }
    if sends.iter().all(|s| s.is_ok()) {
        Ok(sends.into_iter().map(|s| s.unwrap()).collect())
    } else {
        Err(ProducerError::PartialSend(sends).into())
    }
}
```

As you can see in the implementation of `send_all`, it just calls `send` on the messages one by one. The only difference is that it generates `PartialSend` error. Actually, do we need to manually batch the messages and use `send_all`?

Reply:
> in the future we can extract the common logic and avoid so many duplicated code in kafka and pulsar.

Make sense! Then this part maybe future refactored but we just keep it the same at present.